### PR TITLE
BUG: ma.testutils.assert_equal on nans throws AssertionError #6661

### LINKED ISSUE #6661: Bug: ma.testutils.assert_equal on nans throws AssertionError
`np.ma.testutils.assert_equal(np.nan, np.nan)` throws an AssertionError. In my opinion it should no consistent with the behavior of `np.testing.assert_equal(np.nan, np.nan)`.

### PR DESCRIPTION
**Pull request by mcdaniel67**
Fixes #6661 and makes `np.ma.testutils.assert_equal`'s functionality more closely match `np.testing.assert_equal`

Previous Functionality:

```
>>> import numpy as np
>>> from numpy.ma.testutils import assert_equal
>>> from numpy.testing import assert_equal as a_e
>>> a_e(np.nan, np.nan)
>>> assert_equal(np.nan, np.nan)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "numpy/ma/testutils.py", line 130, in assert_equal
    raise AssertionError(msg)
AssertionError:
Items are not equal:
 ACTUAL: nan
 DESIRED: nan
```

Modified Functionality:

```
>>> import numpy as np
>>> from numpy.ma.testutils import assert_equal
>>> from numpy.testing import assert_equal as a_e
>>> a_e(np.nan, np.nan)
>>> assert_equal(np.nan, np.nan)
```

### CODE REVIEW

File: numpy/ma/testutils.py [hunk 1]
```diff
@@ -121,7 +121,11 @@ def assert_equal(actual, desired, err_msg=''):
                 raise AssertionError("%s not in %s" % (k, actual))
             assert_equal(actual[k], desired[k], 'key=%r\n%s' % (k, err_msg))
         return
-    # Case #2: lists .....
+    # Case #2: nan
+    if isinstance(actual, float) and isinstance(desired, float) and \
```
Comment (charris) [inline numpy/ma/testutils.py]: Use `(...)` instead of line continuation here.

**Comment by charris** [comment]
Oh, this is a horrible, horrible function. Let me think about it...

**Comment by mcdaniel67** [comment]
Ok, either way that should be a minor change to make. The reason I elected to go for the line continuation was matching the style used in line 137. And yeah, this function is... interesting.

Would it be worth adding unit tests for `np.ma.testutils.assert_equal` and working on refactoring?

**Comment by charris** [comment]
Note xor with two booleans is `!=`, so that simplifies some stuff also.

Yes, a test would be highly desirable, required even ;)

**Comment by charris** [comment]
Here is equivalent
```
def assert_equal(actual, desired, err_msg=''):
    if not (isinstance(actual, masked_array) or
            isinstance(desired, masked_array)):
        return utils.assert_equal(actual, desired)
    if isinstance(desired, dict):
        raise AssertionError()
    if (actual is masked) != (desired is masked):
        raise ValueError()

    actual = np.asanyarray(actual)
    desired = np.asanyarray(desired)
    if actual.dtype.kind == "S" and desired.dtype.kind == "S":
        actual = actual.tolist()
        desired = actual.tolist()
        return _assert_equal_on_sequences(actual, desired, err_msg=err_msg)
    return assert_array_equal(actual, desired, err_msg)
```
Note that unicode strings are not handled (python 3) and I'm not clear on why the special casing unless there is (was) a problem with fill value. Should be able to use an empty string for masked fill values I think, but currently maskled string arrays use "N/A". We should consider changing that.

**Comment by mcdaniel67** [comment]
Ok, thanks! Should I use that as a refactored base to make my changes on top of?

**Comment by charris** [comment]
It needs some filling out with error messages and such, it is a stripped down version. I think it should solve your problem as is, delegating it to the unmasked version. Writing some tests first would be good. The string problem needs an enhancement. I think we should be using an empty string in those cases.

**Comment by charris** [comment]
I think the strings can be handled like so. `sixu` needs import from `numpy.compat`.
```
def assert_equal(actual, desired, err_msg=''):
    # if nothing masked, default to ndarray version.
    if not (isinstance(actual, masked_array) or
            isinstance(desired, masked_array)):
        return utils.assert_equal(actual, desired)

    # if desired it a dict, actual must be also.
    if isinstance(desired, dict):
        raise AssertionError()
    # masked only compares with masked.
    if (actual is masked) != (desired is masked):
        raise ValueError()

    actual = np.asanyarray(actual)
    desired = np.asanyarray(desired)

    # Deal with string types
    if actual.dtype.kind in "SU":
        val = b'' if actual.dtype.kind == "S" else sixu('')
        actual = masked_array(actual, copy=False, subok=True, fill_value=val)
    if desired.dtype.kind in "SU":
        val = b'' if desired.dtype.kind == "S" else sixu('')
        desired = masked_array(desired, copy=False, subok=True, fill_value=val)

    return assert_array_equal(actual, desired, err_msg)
```

**Comment by mcdaniel67** [comment]
Ah alright, perfect! Thanks for the help, I'll try my hand at implementing the described changes!

**Comment by mcdaniel67** [comment]
Should I push my commits directly to this branch, or would the refactoring indicate that I should do this as a separate PR?

**Comment by charris** [comment]
Just push to you branch. When everything is ready, you can squash your commits together and rewrite the first commit message to match format in http://docs.scipy.org/doc/numpy-1.10.1/dev/gitwash/development_workflow.html, followed by a force push `git push -f origin <your branch>`.

**Comment by mcdaniel67** [comment]
@charris Could you re-review this? I think I've gotten it into a pretty good state.

**Comment by mcdaniel67** [comment]
@charris Should I recreate this PR or will it still be reviewed? I know it's far enough down in the listed PRs now that it has the potential to be skipped over easily.

File: numpy/testing/utils.py [hunk 2]
```diff
@@ -347,6 +347,8 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
     try:
         # isscalar test to check cases such as [np.nan] != np.nan
         if isscalar(desired) != isscalar(actual):
+            if desired == actual:
```
Comment (charris) [inline numpy/testing/utils.py]: Why is this needed?

**Comment by charris** [comment]
I'll need to take a closer look, I note that you did things differently than suggested, so I need to restudy the problem.

File: numpy/ma/testutils.py [hunk 3]
```diff
@@ -91,6 +91,21 @@ def _assert_equal_on_sequences(actual, desired, err_msg=''):
     return
 
 
+def _assert_equal_both_non_masked(actual, desired, err_msg=''):
```
Comment (charris) [inline numpy/ma/testutils.py]: I don't think this should be a separate function, it makes the code harder to follow and is only used once.

File: numpy/ma/testutils.py [hunk 4]
```diff
@@ -91,6 +91,21 @@ def _assert_equal_on_sequences(actual, desired, err_msg=''):
...
+
+    """
+    if (isinstance(desired, (list, tuple)) and
+            isinstance(actual, (list, tuple))):
+        return _assert_equal_on_sequences(actual, desired, err_msg)
+    elif (isinstance(actual, np.matrix) or
+            isinstance(desired, np.matrix)):
+        return assert_array_equal(actual, desired, err_msg)
+    else:
+        return utils.assert_equal(actual, desired, err_msg)
```
Comment (charris) [inline numpy/ma/testutils.py]: I think this is the only function you need. `utils.assert_equal` checks sequences/tuples and handles the matrix case as matrix is a subclass or ndarray.

File: numpy/ma/testutils.py [hunk 5]
```diff
@@ -111,37 +126,22 @@ def assert_equal(actual, desired, err_msg=''):
...
-                            err_msg, header='', names=('x', 'y'))
+    # Delegate assertion if actual isn't masked and desired isn't masked.
+    if not (isinstance(actual, masked_array) or
+            isinstance(desired, masked_array)):
+        return _assert_equal_both_non_masked(actual, desired, err_msg)
+
+    # Masked only compares with masked.
+    if (actual is masked) != (desired is masked):
+        msg = build_err_msg([actual, desired], err_msg, header='', names=('x', 'y'))
         raise ValueError(msg)
```
Comment (charris) [inline numpy/ma/testutils.py]: I wonder if we should change this to raise `AssertionError`? If two objects can't be compared they can't be equal. I wonder why this this was done?

File: numpy/ma/testutils.py [hunk 6]
```diff
@@ -111,37 +126,22 @@ def assert_equal(actual, desired, err_msg=''):
...
-    (actual_dtype, desired_dtype) = (actual.dtype, desired.dtype)
-    if actual_dtype.char == "S" and desired_dtype.char == "S":
-        return _assert_equal_on_sequences(actual.tolist(),
-                                          desired.tolist(),
-                                          err_msg='')
+
+    actual = np.asanyarray(actual)
+    desired = np.asanyarray(desired)
+    if actual.dtype.char == "S" and desired.dtype.char == "S":
+        actual = actual.tolist()
```
Comment (charris) [inline numpy/ma/testutils.py]: This still looks wrong to me. Note that I suggested a view in order to change the fill value to `''` and then call `assert_array_equal` (which fills the masked values).

**Comment by charris** [comment]
Close and reopen to see if tests of unsupported Python versions go away. Otherwise may need a rebase on master

**Comment by charris** [comment]
@mcdaniel67 Needs to be finished.
